Disallow dynamic cast to extern(C++) class

ntrel opened the pull request:
See Issue 21690 - Unable to dynamic cast extern(C++) classes.

Instead, use `cast(Derived)cast(void*)base`.

ntrel committed:
Disallow dynamic cast on extern(C++) class

See Issue 21690 - Unable to dynamic cast extern(C++) classes.

SixthDot commented:
That should be deprecated first. It's not that that this does not work, it's more that the expected effect is not encountered. Problem being that D CastExp does plenty of different casts depending on the source and the target.

SixthDot reviewed (changes requested):
That should be deprecated first. It's not that that this does not work, it's more that the expected effect is not encountered. Problem being that D CastExp does plenty of different casts depending on the source and the target.

Bolpat commented:
What does this PR do exactly? If it makes casting a C++ class impossible, there is no way to fix code. There must be a way to `static cast` C++ class objects.

ntrel committed:
Fix segfault

ntrel committed:
Make deprecation not error

ntrel commented:
@Bolpat This deprecates casting any class instance to an extern(C++) class type. If the source expression is not a class it still works.

ntrel committed:
Allow cast to base class

ntrel committed:
Add workaround for cast to derived class test

SixthDot reviewed (approved):
with the deprecation this goes in the right direction.

ibuclaw commented on compiler/src/dmd/expressionsem.d:
```diff
@@ -7825,6 +7825,18 @@ private extern (C++) final class ExpressionSemanticVisitor : Visitor
+            if (e1c && toc.classKind == ClassKind.cpp &&
+                !toc.isBaseOf(e1c, null))
+            {
+                // @@@DEPRECATED_2.114@@@
```
It'll be 2.115 now to make it the full 10.

ntrel committed:
Only deprecate downcasts & show workaround

ntrel committed:
another test workaround

ntrel committed:
update date, `deprecation` call

ntrel committed:
Fix Issue 23957 - Casting to derived extern(C++) class is unsafe

WalterBright commented on compiler/src/dmd/expressionsem.d:
```diff
@@ -7825,6 +7825,19 @@ private extern (C++) final class ExpressionSemanticVisitor : Visitor
+                // @@@DEPRECATED_2.115@@@
+                // When turning into error, uncomment the return statement
+                exp.deprecation("dynamic cast is not implemented for `extern(C++)` classes");
+                deprecationSupplemental(exp.loc, "use `cast(T)cast(void*)obj` instead");
```
ag0aep6g writes:

`cast(D*) &c` is the unambiguous type paint. `cast(D) cast(void*) c` can be intercepted by `opCast`.

ntrel commented:
I haven't worked on finishing this as it seems quite disruptive given the NG thread about deprecating code that is working properly (even if it catches some code that isn't working). Is there a plan for changes like this - should they be delayed to a special release, or triggered by a preview switch instead?

dkorpel commented:
> Is there a plan for changes like this

It's still being discussed, but I think this PR is fine, since there's a good reason and a clear migration path.

RazvanN7 commented on compiler/src/dmd/expressionsem.d:
```diff
@@ -7813,6 +7813,18 @@ private extern (C++) final class ExpressionSemanticVisitor : Visitor
                 return setError();
         }
+        if (global.compileEnv.obsolete)
+        if (auto toc = exp.to.isClassHandle())
```
This style is kind of weird. Is this intended? (if's having the same indentation).

WalterBright reviewed (changes requested):
I approve if the test suite failures are addressed.

ntrel committed:
Change suggestion to avoid `opCast!(void*)`

`&` requires an lvalue, so not ideal, but at least it it always works when
using an lvalue.

thewilsonator commented:
(please retitle the commit that fixes the issue "Fix bugzilla issue 23957 - Casting to derived extern(C++) class is unsafe")

thewilsonator commented:
See also the deprecations:
```
src/dmd/asttypename.d(58): Deprecation: dynamic cast is not implemented for `extern(C++)` classes
src/dmd/asttypename.d(58):        use `*cast(dmd.expression.Expression*) &object` instead
src/dmd/asttypename.d(60): Deprecation: dynamic cast is not implemented for `extern(C++)` classes
src/dmd/asttypename.d(60):        use `*cast(dmd.dsymbol.Dsymbol*) &object` instead
src/dmd/asttypename.d(62): Deprecation: dynamic cast is not implemented for `extern(C++)` classes
src/dmd/asttypename.d(62):        use `*cast(dmd.mtype.Type*) &object` instead
src/dmd/asttypename.d(64): Deprecation: dynamic cast is not implemented for `extern(C++)` classes
src/dmd/asttypename.d(64):        use `*cast(dmd.mtype.Parameter*) &object` instead
src/dmd/asttypename.d(66): Deprecation: dynamic cast is not implemented for `extern(C++)` classes
src/dmd/asttypename.d(66):        use `*cast(dmd.statement.Statement*) &object` instead
src/dmd/asttypename.d(68): Deprecation: dynamic cast is not implemented for `extern(C++)` classes
src/dmd/asttypename.d(68):        use `*cast(dmd.cond.Condition*) &object` instead
src/dmd/asttypename.d(70): Deprecation: dynamic cast is not implemented for `extern(C++)` classes
src/dmd/asttypename.d(70):        use `*cast(dmd.dtemplate.TemplateParameter*) &object` instead
src/dmd/asttypename.d(72): Deprecation: dynamic cast is not implemented for `extern(C++)` classes
src/dmd/asttypename.d(72):        use `*cast(dmd.init.Initializer*) &object` instead
src/dmd/clone.d(60): Deprecation: dynamic cast is not implemented for `extern(C++)` classes
src/dmd/clone.d(60):        use `*cast(dmd.mtype.TypeFunction*) &object` instead
src/dmd/clone.d(180): Deprecation: dynamic cast is not implemented for `extern(C++)` classes
src/dmd/clone.d(180):        use `*cast(dmd.mtype.TypeStruct*) &object` instead
src/dmd/clone.d(286): Deprecation: dynamic cast is not implemented for `extern(C++)` classes
src/dmd/clone.d(286):        use `*cast(dmd.mtype.TypeStruct*) &object` instead
src/dmd/clone.d(317): Deprecation: dynamic cast is not implemented for `extern(C++)` classes
src/dmd/clone.d(317):        use `*cast(dmd.mtype.TypeFunction*) &object` instead
src/dmd/clone.d(438): Deprecation: dynamic cast is not implemented for `extern(C++)` classes
src/dmd/clone.d(438):        use `*cast(dmd.mtype.TypeStruct*) &object` instead
src/dmd/clone.d(566): Deprecation: dynamic cast is not implemented for `extern(C++)` classes
src/dmd/clone.d(566):        use `*cast(dmd.mtype.TypeFunction*) &object` instead
src/dmd/clone.d(641): Deprecation: dynamic cast is not implemented for `extern(C++)` classes
src/dmd/clone.d(641):        use `*cast(dmd.mtype.TypeFunction*) &object` instead
src/dmd/clone.d(765): Deprecation: dynamic cast is not implemented for `extern(C++)` classes
src/dmd/clone.d(765):        use `*cast(dmd.mtype.TypeStruct*) &object` instead
src/dmd/clone.d(806): Deprecation: dynamic cast is not implemented for `extern(C++)` classes
src/dmd/clone.d(806):        use `*cast(dmd.mtype.TypeFunction*) &object` instead
src/dmd/clone.d(906): Deprecation: dynamic cast is not implemented for `extern(C++)` classes
src/dmd/clone.d(906):        use `*cast(dmd.mtype.TypeStruct*) &object` instead
src/dmd/clone.d(1268): Deprecation: dynamic cast is not implemented for `extern(C++)` classes
src/dmd/clone.d(1268):        use `*cast(dmd.mtype.TypeStruct*) &object` instead
src/dmd/clone.d(1278): Deprecation: dynamic cast is not implemented for `extern(C++)` classes
```

ntrel commented:
> I think this PR is fine, since there's a good reason and a clear migration path.

Given that people are using these casts expecting them to work as C++ `static_cast` in working D code (including dmd), wouldn't it be better to specify that down casting `extern(C++)` classes is a type paint? Then we only need #15294 to make those not `@safe`.

It's also not ideal that the compiler has to recommend the type paint which only works with lvalue objects - so fixing rvalue expression casts is more awkward (though in practice, `cast(Derived)cast(void*) rvalueExpression` is often fine).
